File: examples/android-demo/src/main/java/com/tencent/mtt/hippy/example/module/turbo/TurboArray.java
package com.tencent.mtt.hippy.example.module.turbo;

import com.tencent.mtt.hippy.annotation.HippyTurboObj;
import com.tencent.mtt.hippy.annotation.HippyTurboProp;
import com.tencent.mtt.hippy.common.HippyArray;
import com.tencent.mtt.hippy.common.HippyMap;

import java.util.ArrayList;

/**
 * 支持以引用的方式暴露给js使用的Array
 * TurboArray支持的类型如下：
 * 【框架转换后透传】int/long/double/String/boolean/HippyMap/HippyArray
 * 【引用】TurboArray/TurboMap
 */
@HippyTurboObj
public class TurboArray {

  private final ArrayList mDatas;

  public TurboArray() {
    mDatas = new ArrayList();
  }

  @HippyTurboProp(expose = true)
  public int size() {
    return mDatas.size();
  }

  @HippyTurboProp(expose = true)
  public int getInt(int index) {
    if (mDatas.size() > index) {
      Object value = mDatas.get(index);
      return value instanceof Number ? ((Number) value).intValue() : 0;
    }
    return 0;
  }

  @HippyTurboProp(expose = true)
  public long getLong(int index) {
    if (mDatas.size() > index) {
      Object value = mDatas.get(index);
      return value instanceof Number ? ((Number) value).longValue() : 0;
    }
    return 0;
  }

  @HippyTurboProp(expose = true)
  public double getDouble(int index) {
    if (mDatas.size() > index) {
      Object value = mDatas.get(index);
      return value instanceof Number ? ((Number) value).doubleValue() : 0;
    }
    return 0;
  }

  @HippyTurboProp(expose = true)
  public String getString(int index) {
    if (mDatas.size() > index) {
      return String.valueOf(mDatas.get(index));
    }
    return null;
  }

  @HippyTurboProp(expose = true)
  public boolean getBoolean(int index) {
    if (mDatas.size() > index) {
      Object value = mDatas.get(index);
      return (value instanceof Boolean) && (boolean) value;
    }
    return false;
  }

  @HippyTurboProp(expose = true)
  public TurboArray getTurboArray(int index) {
    if (mDatas.size() > index) {
      Object value = mDatas.get(index);
      return value instanceof TurboArray ? (TurboArray) value : null;
    }
    return null;
  }

  @HippyTurboProp(expose = true)
  public TurboMap getTurboMap(int index) {
    if (mDatas.size() > index) {
      Object value = mDatas.get(index);
      return value instanceof TurboMap ? (TurboMap) value : null;
    }
    return null;
  }

  @HippyTurboProp(expose = true)
  public HippyArray getHippyArray(int index) {
    if (mDatas.size() > index) {
      Object value = mDatas.get(index);
      return value instanceof HippyArray ? (HippyArray) value : null;
    }
    return null;
  }

  @HippyTurboProp(expose = true)
  public HippyMap getHippyMap(int index) {
    if (mDatas.size() > index) {
      Object value = mDatas.get(index);
      return value instanceof HippyMap ? (HippyMap) value : null;
    }
    return null;
  }

  public void pushInt(int value) {
    mDatas.add(value);
  }

  public void pushLong(long value) {
    mDatas.add(value);
  }

  public void pushDouble(double value) {
    mDatas.add(value);
  }

  public void pushBoolean(boolean value) {
    mDatas.add(value);
  }

  public void pushString(String value) {
    mDatas.add(value);
  }

  public void pushTurboArray(TurboArray array) {
    mDatas.add(array);
  }

  public void pushTurboMap(TurboMap map) {
    mDatas.add(map);
  }

  public void pushHippyArray(HippyArray array) {
    mDatas.add(array);
  }

  public void pushHippyMap(HippyMap map) {
    mDatas.add(map);
  }

  public void pushNull() {
    mDatas.add(null);
  }

  public void clear() {
    mDatas.clear();
  }

  @Override
  public String toString() {
    return mDatas.toString();
  }

}

File: framework/examples/android-demo/src/main/java/com/openhippy/example/turbo/TurboArray.kt
/*
 * Tencent is pleased to support the open source community by making Hippy
 * available.
 * Copyright (C) 2018 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.openhippy.example.turbo

import com.tencent.mtt.hippy.annotation.HippyTurboObj
import com.tencent.mtt.hippy.annotation.HippyTurboProp
import com.tencent.mtt.hippy.common.HippyArray
import com.tencent.mtt.hippy.common.HippyMap

/**
 * An `Array` type that can be used as reference by Js
 *
 * The following data types are supported:
 * - **As primitive**: int/long/double/String/boolean/HippyMap/HippyArray
 * - **As reference**: TurboArray/TurboMap
 */
@HippyTurboObj
class TurboArray {
    private val mData: ArrayList<Any?> = ArrayList()

    @HippyTurboProp(expose = true)
    fun size(): Int = mData.size

    @HippyTurboProp(expose = true)
    fun getInt(index: Int): Int = (mData.getOrNull(index) as? Number)?.toInt() ?: 0

    @HippyTurboProp(expose = true)
    fun getLong(index: Int): Long = (mData.getOrNull(index) as? Number)?.toLong() ?: 0

    @HippyTurboProp(expose = true)
    fun getDouble(index: Int): Double = (mData.getOrNull(index) as? Number)?.toDouble() ?: 0.0

    @HippyTurboProp(expose = true)
    fun getString(index: Int): String? = mData.getOrNull(index)?.toString()

    @HippyTurboProp(expose = true)
    fun getBoolean(index: Int): Boolean = (mData.getOrNull(index) as? Boolean) ?: false

    @HippyTurboProp(expose = true)
    fun getTurboArray(index: Int): TurboArray? = mData.getOrNull(index) as? TurboArray

    @HippyTurboProp(expose = true)
    fun getTurboMap(index: Int): TurboMap? = mData.getOrNull(index) as? TurboMap

    @HippyTurboProp(expose = true)
    fun getHippyArray(index: Int): HippyArray? = mData.getOrNull(index) as? HippyArray

    @HippyTurboProp(expose = true)
    fun getHippyMap(index: Int): HippyMap? = mData.getOrNull(index) as? HippyMap

    fun pushInt(value: Int) {
        mData.add(value)
    }

    fun pushLong(value: Long) {
        mData.add(value)
    }

    fun pushDouble(value: Double) {
        mData.add(value)
    }

    fun pushBoolean(value: Boolean) {
        mData.add(value)
    }

    fun pushString(value: String?) {
        mData.add(value)
    }

    fun pushTurboArray(array: TurboArray?) {
        mData.add(array)
    }

    fun pushTurboMap(map: TurboMap?) {
        mData.add(map)
    }

    fun pushHippyArray(array: HippyArray?) {
        mData.add(array)
    }

    fun pushHippyMap(map: HippyMap?) {
        mData.add(map)
    }

    fun pushNull() {
        mData.add(null)
    }

    fun clear() {
        mData.clear()
    }

    override fun toString(): String = mData.toString()
}